trident/ion_balance.py

```diff
@@ -686,23 +686,50 @@ def add_ion_mass_field(atom, ion, ds, ftype="gas",
                sampling_type=sampling_type)
     if ion == 1: # add aliased field too
         _alias_field(ds, ("gas", alias_field), ("gas", field))
-
+        
 def _ion_mass(field, data):
     """
-    Creates the function for a derived field to follow the total mass of an
-    ion over a dataset given that the specified ion's density field exists
-    in the dataset.
+    Creates the function for a derived field for following the mass
+    of an ion over a dataset given that the specified ion's ion_fraction field
+    exists in the dataset.
     """
     if isinstance(field.name, tuple):
         ftype = field.name[0]
         field_name = field.name[1]
     else:
         ftype = "gas"
         field_name = field.name
+    atom = field_name.split("_")[0]
     prefix = field_name.split("_mass")[0]
     suffix = field_name.split("_mass")[-1]
     fraction_field_name = "%s_ion_fraction%s" % (prefix, suffix)
-    return data[ftype, fraction_field_name] * data["gas", "mass"]
+
+    # try the atom-specific density field first
+    nuclei_field = "%s_nuclei_mass_density" % atom
+    if (ftype, nuclei_field) in data.ds.field_info:
+        return data[fraction_field_name] * \
```

Comment:
I think this just needs `ftype`.
```suggestion
        return data[ftype, fraction_field_name] * \
```